services-camel/src/main/java/edu/unc/lib/dl/services/camel/fulltext/FulltextProcessor.java

```diff
@@ -86,30 +76,16 @@ public void process(Exchange exchange) throws Exception {
             return;
         }
 
-        int retryAttempt = 0;
+        Path derivativePath = Paths.get(derivativeBasePath, binarySubPath + "/" + binaryId + ".txt");
+        File derivative = derivativePath.toFile();
+        File parentDir = derivative.getParentFile();
 
-        InputStream binaryStream = new ByteArrayInputStream(text.getBytes(StandardCharsets.UTF_8));
-
-        while (true) {
-            try {
-                BinaryObject binary = repoObjLoader.getBinaryObject(PIDs.get(binaryUri));
-                FileObject parent = (FileObject) binary.getParent();
-
-                parent.addDerivative(slug, binaryStream, fileName, MIMETYPE, PcdmUse.ExtractedText);
-
-                log.info("Adding derivative for {} from {}", binaryUri, fileName);
-                break;
-            } catch (Exception e) {
-                if (retryAttempt == maxRetries) {
-                    throw e;
-                }
-
-                retryAttempt++;
-                log.info("Unable to add derivative for {} from {}. Retrying, attempt {}",
-                        binaryUri, binaryPath, retryAttempt);
-                TimeUnit.MILLISECONDS.sleep(retryDelay);
+        if (parentDir != null) {
+            parentDir.mkdirs();
+        }
 
-            }
+        try (PrintWriter fulltext = new PrintWriter(derivativePath.toString())) {
+            fulltext.println(text);
```

Comment:
i'm concerned that if the text extraction above fails for any reason other than a tika exception we have a NPE on our hands here at line 88

Reply:
I think most likely issues in here will throw IOExceptions. I have updated FullTextProcessor and AddDerivativesProcessor to now throw IOExceptions if they are unable to create parent directories for the final derivative instead of throwing a file not found exception a few lines down.